src/inject.cc

```diff
@@ -8,68 +8,119 @@
 #include <vector>
 
-bool InjectDll(const uint32_t ProcessId, const std::filesystem::path &Path) {
+
+HANDLE OpenRemoteProcess(const uint32_t ProcessId) {
   const uint32_t ProcessRights =
       PROCESS_CREATE_THREAD | PROCESS_QUERY_INFORMATION | PROCESS_VM_OPERATION |
       PROCESS_VM_WRITE | PROCESS_VM_READ;
   const HANDLE Process = OpenProcess(ProcessRights, false, ProcessId);
 
   if (Process == nullptr) {
-    return EXIT_FAILURE;
+    printf("Failed to open the remote process.\n");
+    return INVALID_HANDLE_VALUE;
   }
 
-  const PVOID RemoteDllPath = VirtualAllocEx(
-      Process, nullptr, 0x1000, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+  return Process;
+}
+
+PVOID AllocateMemInRemoteProcess(HANDLE hRemoteProcess,
+                                 const std::filesystem::path &Path) {
+  const std::string DllPath = Path.string();
+  const size_t DllPathLen = DllPath.size() + 1;
+  SIZE_T BytesWritten;
+
+  const PVOID RemoteDllPath = VirtualAllocEx(hRemoteProcess, nullptr, 0x1000, MEM_COMMIT | MEM_RESERVE,
+                     PAGE_READWRITE);
 
   if (RemoteDllPath == nullptr) {
     printf("VirtualAllocEx failed.\n");
-    return EXIT_FAILURE;
+    goto Error;
   }
 
-  const std::string DllPath = Path.string();
-  const size_t DllPathLen = DllPath.size() + 1;
-  SIZE_T BytesWritten;
-  if (!WriteProcessMemory(Process, RemoteDllPath, DllPath.c_str(), DllPathLen,
+  if (!WriteProcessMemory(hRemoteProcess, RemoteDllPath, DllPath.c_str(),
+                          DllPathLen,
                           &BytesWritten)) {
+    VirtualFreeEx(hRemoteProcess, RemoteDllPath, 0, MEM_RELEASE);
     printf("WriteProcessMemory failed.\n");
-    return EXIT_FAILURE;
+    goto Error;
   }
 
+  return RemoteDllPath;
+
+Error:
+  CloseHandle(hRemoteProcess);
+  return nullptr;
+}
+
+HANDLE CreateThreadInRemoteProcess(HANDLE hRemoteProcess, PVOID RemoteDllPath) {
+  PVOID LoadLibraryA;
+  HANDLE Thread;
+  DWORD Tid;
+
   const HMODULE Kernelbase = GetModuleHandleA("kernelbase");
   if (Kernelbase == nullptr) {
     printf("GetModuleHandleA failed.\n");
-    return EXIT_FAILURE;
+    goto Error;
   }
 
-  const PVOID LoadLibraryA = PVOID(GetProcAddress(Kernelbase, "LoadLibraryA"));
+  LoadLibraryA = PVOID(GetProcAddress(Kernelbase, "LoadLibraryA"));
   if (LoadLibraryA == nullptr) {
     printf("GetProcAddress failed.\n");
-    return EXIT_FAILURE;
+    goto Error;
   }
 
-  DWORD Tid;
-  const HANDLE Thread = CreateRemoteThread(Process, nullptr, 0,
+  Thread = CreateRemoteThread(hRemoteProcess, nullptr, 0,
                                            LPTHREAD_START_ROUTINE(LoadLibraryA),
                                            RemoteDllPath, 0, &Tid);
 
   if (Thread == NULL) {
     printf("CreateRemoteThread failed.\n");
-    return EXIT_FAILURE;
+    goto Error;
+  }
+
+  printf("Thread with ID %d has been created.\n", Tid);
+
+  return Thread;
+
+Error:
+  VirtualFreeEx(hRemoteProcess, RemoteDllPath, 0, MEM_RELEASE);
```

Comment:
I think this cleanup is a bit confusing - as it also exists in the caller of this function; shouldn't we free the remote mem in the caller function only?